Validation/HGCalTriggerPrimitives/plugins/HGCalTriggerValidator.cc

```diff
@@ -0,0 +1,281 @@
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Common/interface/Provenance.h"
+
+#include <DQMServices/Core/interface/DQMStore.h>
+#include <DQMServices/Core/interface/MonitorElement.h>
+#include <DQMServices/Core/interface/DQMEDAnalyzer.h>
+
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
+#include "DataFormats/L1THGCal/interface/HGCalTower.h"
+
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+
+#include "L1Trigger/L1THGCal/interface/backend/HGCalTriggerClusterIdentificationBase.h"
+
+//
+// class declaration
+//
+
+class HGCalTriggerValidator : public DQMEDAnalyzer {
+
+  public:
+    explicit HGCalTriggerValidator(const edm::ParameterSet&);
+    ~HGCalTriggerValidator() override;
+
+    void analyze(const edm::Event&, const edm::EventSetup&) override;
+
+  protected:
+    void bookHistograms(DQMStore::IBooker &, edm::Run const &, edm::EventSetup const &) override;
+
+  private:
+// ----------member data ---------------------------
+    //histogram tc related
+    MonitorElement *h_tc_n_, *h_tc_layer_, *h_tc_mipPt_;
+    MonitorElement *h_tc_pt_, *h_tc_x_, *h_tc_y_; 
+    MonitorElement *h_tc_z_, *h_tc_energy_, *h_tc_eta_, *h_tc_phi_;
+
+    //histogram cl related
+    MonitorElement *h_cl_n_, *h_cl_mipPt_, *h_cl_pt_; 
+    MonitorElement *h_cl_energy_, *h_cl_eta_, *h_cl_phi_;
+    MonitorElement *h_cl_layer_, *h_cl_cells_n_;
+
+    //histogram multicl related
+    MonitorElement *h_cl3d_n_, *h_cl3d_pt_, *h_cl3d_energy_;
+    MonitorElement *h_cl3d_eta_, *h_cl3d_phi_, *h_cl3d_clusters_n_;
+    // cluster shower shapes
+    MonitorElement *h_cl3d_showerlength_, *h_cl3d_coreshowerlength_, *h_cl3d_firstlayer_;
+    MonitorElement *h_cl3d_maxlayer_, *h_cl3d_seetot_, *h_cl3d_seemax_;
+    MonitorElement *h_cl3d_spptot_, *h_cl3d_sppmax_, *h_cl3d_szz_;
+    MonitorElement *h_cl3d_srrtot_, *h_cl3d_srrmax_, *h_cl3d_srrmean_;
+    MonitorElement *h_cl3d_emaxe_, *h_cl3d_bdteg_, *h_cl3d_quality_;
+
+    //histogram tower related
+    MonitorElement *h_tower_n_, *h_tower_pt_, *h_tower_energy_;
+    MonitorElement *h_tower_eta_, *h_tower_phi_, *h_tower_etEm_;
+    MonitorElement *h_tower_etHad_, *h_tower_iEta_, *h_tower_iPhi_;
+
+
+    edm::EDGetToken trigger_cells_token_;
+    edm::EDGetToken clusters_token_;
+    edm::EDGetToken multiclusters_token_; 
+    edm::EDGetToken towers_token_;
+   
+   std::unique_ptr<HGCalTriggerClusterIdentificationBase> id_;
+   
+private:
+    int tc_n_;
+    int cl_n_;
+    int cl3d_n_;
+    int tower_n_;
```

Comment:
In principle, these four variables (`tc_n_`, `cl_n_`, `cl3d_n_`, `tower_n_`) don't need to be data members. Local variables would be enough.